### ISSUE TITLE
Build fails when versioning.enabled = false

### ISSUE THREAD

**Issue by mirkoscherf** [DSpace/DSpace#8985]
**Describe the bug**
A clear and concise description of what the bug is.  Include the version(s) of DSpace where you've seen this problem. Link to examples if they are public.
DSpace 7.6
Building backend with disabled item versioning causes build to fail due to missing HandleIdentifierProvider in identifier-service.xml.

**To Reproduce**
1. set `versioning.enabled = false` in versioning.cfg
2. (optional) in identifier-service.xml comment out bean with id="org.dspace.identifier.HandleIdentifierProvider"
3. build backend

**Expected behavior**
The build should succeed.

**Related work**

**Comment by mirkoscherf** [comment]
I will provide a PR.

**Comment by mirkoscherf** [comment]
@tdonohue I'm not quite sure how to proceed with the unit tests. Integration tests run through just fine but the unit tests rely on versioning being enabled, so they fail miserably.
`mvn install -DskipUnitTests=false` fails and `mvn test -DskipUnitTests=false '-Dtest=org.dspace.identifier.*Test'` does not execute any tests.

**Comment by tdonohue** [comment]
@mirkoscherf : I'd recommend creating a draft PR with the changes you've made so that others can more easily see the failing unit test behavior.  It's difficult to help without understanding/seeing the errors.  It's possible that some individual tests need to have `versioning.enabled=true`, but in that scenario those tests could set that value in the test itself (see `ConfigurationService.setProperty()` and how it's used in other tests).  But, it's unclear to me which tests you reference, so it's difficult to provide advice.

**Comment by mirkoscherf** [comment]
PR is underway.
The failed tests look like this:
```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.607 s <<< FAILURE! - in org.dspace.xmlworkflow.XmlWorkflowFactoryTest
[ERROR] org.dspace.xmlworkflow.XmlWorkflowFactoryTest  Time elapsed: 0.607 s  <<< ERROR!
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'org.dspace.app.sherpa.submit.SHERPASubmitService' defined in file [F:\workspace\dspace7\src\dspace\dspace-api\target\testing\dspace\config\spring\api\sherpa.xml]: Cannot resolve reference to bean 'org.dspace.app.sherpa.submit.SHERPASubmitConfigurationService' while setting bean property 'configuration'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'org.dspace.app.sherpa.submit.SHERPASubmitConfigurationService' defined in file [F:\workspace\dspace7\src\dspace\dspace-api\target\testing\dspace\config\spring\api\sherpa.xml]: Cannot create inner bean 'org.dspace.app.sherpa.submit.MetadataValueISSNExtractor#743997e0' of type [org.dspace.app.sherpa.submit.MetadataValueISSNExtractor] while setting bean property 'issnItemExtractors' with key [0]; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating 
bean with name 'org.dspace.app.sherpa.submit.MetadataValueISSNExtractor#743997e0': Unsatisfied dependency expressed through field 'itemService'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.dspace.content.ItemServiceImpl#0': Unsatisfied dependency expressed through field 'handleService'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.dspace.handle.HandleServiceImpl#0': Unsatisfied dependency expressed through field 'siteService'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.dspace.content.SiteServiceImpl#0': Unsatisfied dependency expressed through field 'metadataValueService'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.dspace.content.MetadataValueServiceImpl#0': Unsatisfied dependency 
expressed through field 'authorizeService'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.dspace.authorize.AuthorizeServiceImpl#0': Unsatisfied dependency expressed through field 'bitstreamService'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.dspace.content.BitstreamServiceImpl#0': Unsatisfied dependency expressed through field 'relationshipService'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.dspace.content.RelationshipServiceImpl#0': Unsatisfied dependency expressed through field 'entityTypeService'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.dspace.content.EntityTypeServiceImpl#0': Unsatisfied dependency expressed through field 'groupService'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.dspace.eperson.GroupServiceImpl#0': Unsatisfied dependency expressed through field 'collectionService'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.dspace.content.CollectionServiceImpl#0': Unsatisfied dependency expressed through field 'resourcePolicyService'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.dspace.authorize.ResourcePolicyServiceImpl#0': Unsatisfied dependency expressed through field 'contentServiceFactory'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'contentServiceFactory': Unsatisfied dependency expressed through field 'dSpaceObjectServices'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'org.dspace.content.CommunityServiceImpl#0': Unsatisfied dependency expressed through field 'identifierService'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'org.dspace.identifier.service.IdentifierService' defined in file [F:\workspace\dspace7\src\dspace\dspace-api\target\testing\dspace\config\spring\api\identifier-service.xml]: Cannot resolve reference to bean 'org.dspace.identifier.HandleIdentifierProvider' while setting bean property 'providers' with key [0]; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'org.dspace.identifier.HandleIdentifierProvider' defined in file [...\src\dspace\dspace-api\target\testing\dspace\config\spring\api\identifier-service.xml]: Invocation of init method failed; 

nested exception is java.lang.RuntimeException: the org.dspace.identifier.VersionedHandleIdentifierProvider is enabled, but the versioning is disabled.
```
Current test run result 
`[ERROR] Tests run: 337, Failures: 0, Errors: 76, Skipped: 1`

**Comment by tdonohue** [comment]
@mirkoscherf : Oh, I see.  Yes, that error is because the tests enable the "VersionedHandleIdentifierProvider" by default (which requires `versioning.enabled=true`).

The only way to solve this would be to:
* Modify tests to always set `versioning.enabled=true` (regardless of your setting in dspace.cfg or local.cfg).  This would be done by adding the setting to the Test Environment local.cfg here: https://github.com/DSpace/DSpace/blob/main/dspace-api/src/test/data/dspaceFolder/config/local.cfg
* OR, locally, change your test environment to use the `HandleIdentifierProvider`.  That'd be done by modifying the `identifier-service.xml` used by the tests here: https://github.com/DSpace/DSpace/blob/main/dspace-api/src/test/data/dspaceFolder/config/spring/api/identifier-service.xml
    * If you choose this option, likely any tests which test for versioning will fail.  They'd need to be disabled individually.

So, basically, because these two Handle providers cannot both be enabled at the same time, the tests need to choose *one*.  Currently, by default they use `VersionedHandleIdentifierProvider` because there are tests that verify that versioning works properly.

Ideally, longer term, it'd obviously be better if the correct "HandleIdentifierProvider" was just *selected* based on the value of `versioning.enabled`.  But, I'm honestly not sure how easy that'd be to implement.

**Comment by tdonohue** [comment]
I believe part of the issue described in this ticket was a documentation error.  The Item Versioning docs were incorrect for how to properly **disable** versioning.  They have been corrected to note there are three steps:

1. Set `versioning.enabled = false`
2. Comment out the `VersionedHandleIdentifierProvider` in `identifier-service.xml` but ALSO REPLACE it with `HandleIdentifierProvider`
3. Removing `versioning` from list of default consumers

See docs at https://wiki.lyrasis.org/display/DSDOC7x/Item+Level+Versioning#ItemLevelVersioning-DisablingItemLevelVersioning

**Comment by nwoodward** [comment]
I followed the steps to disable versioning, including adding the HandleIdentifyProvider to [dspace]/config/spring/api/identifier-services.xml and removing versioning from the default.consumers list. But now I can't create new communities. The stack trace indicates there's an issue with casting in the HandleIdentityProvider.

> 2023-10-13 09:31:57,409 ERROR 2e874867-250d-4b44-8a71-12a185f9189b 3fe9dc00-0a90-4625-bce3-66ff24bd541e org.dspace.app.rest.exception.DSpaceApiExceptionControllerAdvice @ An exception has occurred (status:500)
java.lang.ClassCastException: class org.dspace.content.Community cannot be cast to class org.dspace.content.Item (org.dspace.content.Community and org.dspace.content.Item are in unnamed module of loader org.apache.catalina.loader.ParallelWebappClassLoader @340da44c)
        at org.dspace.identifier.HandleIdentifierProvider.register(HandleIdentifierProvider.java:73) ~[dspace-api-7.6.1-SNAPSHOT.jar:7.6.1-SNAPSHOT]
        at org.dspace.identifier.IdentifierServiceImpl.register(IdentifierServiceImpl.java:107) ~[dspace-api-7.6.1-SNAPSHOT.jar:7.6.1-SNAPSHOT]
        at org.dspace.content.CommunityServiceImpl.create(CommunityServiceImpl.java:128) ~[dspace-api-7.6.1-SNAPSHOT.jar:7.6.1-SNAPSHOT]
        at org.dspace.content.CommunityServiceImpl.create(CommunityServiceImpl.java:93) ~[dspace-api-7.6.1-SNAPSHOT.jar:7.6.1-SNAPSHOT]
        at org.dspace.content.CommunityServiceImpl.create(CommunityServiceImpl.java:88) ~[dspace-api-7.6.1-SNAPSHOT.jar:7.6.1-SNAPSHOT]
        at org.dspace.app.rest.repository.CommunityRestRepository.createCommunity(CommunityRestRepository.java:142) ~[classes/:7.6.1-SNAPSHOT]
        at org.dspace.app.rest.repository.CommunityRestRepository.createAndReturn(CommunityRestRepository.java:123) ~[classes/:7.6.1-SNAPSHOT]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344) ~[spring-aop-5.3.27.jar:5.3.27]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198) ~[spring-aop-5.3.27.jar:5.3.27]
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:799) ~[spring-aop-5.3.27.jar:5.3.27]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-5.3.27.jar:5.3.27]
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763) ~[spring-aop-5.3.27.jar:5.3.27]
        at org.springframework.security.access.intercept.aopalliance.MethodSecurityInterceptor.invoke(MethodSecurityInterceptor.java:61) ~[spring-security-core-5.7.8.jar:5.7.8]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.3.27.jar:5.3.27]
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763) ~[spring-aop-5.3.27.jar:5.3.27]
        at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:708) ~[spring-aop-5.3.27.jar:5.3.27]
        at org.dspace.app.rest.repository.CommunityRestRepository$$EnhancerBySpringCGLIB$$85138c3b.createAndReturn(<generated>) ~[classes/:7.6.1-SNAPSHOT]
        at org.dspace.app.rest.repository.DSpaceRestRepository.createAndReturn(DSpaceRestRepository.java:326) ~[classes/:7.6.1-SNAPSHOT]
        at org.dspace.app.rest.repository.DSpaceRestRepository$$FastClassBySpringCGLIB$$615525f9.invoke(<generated>) ~[classes/:7.6.1-SNAPSHOT]
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218) ~[spring-core-5.3.27.jar:5.3.27]
        at org.springframework.aop.framework.CglibAopProxy.invokeMethod(CglibAopProxy.java:386) ~[spring-aop-5.3.27.jar:5.3.27]
        at org.springframework.aop.framework.CglibAopProxy.access$000(CglibAopProxy.java:85) ~[spring-aop-5.3.27.jar:5.3.27]
...

**Comment by tdonohue** [comment]
@nwoodward : That sounds like this bug / fix: #9098.   It should now be fixed on the latest `dspace-7_x` code and `main` code, but it was only fixed in the last week.

**Comment by nwoodward** [comment]
@tdonohue Thanks! That's exactly what I needed. I wish I had seen that PR come in before disabling versioning on 14 DSpace repositories, but I guess this is what Friday afternoons are for.
